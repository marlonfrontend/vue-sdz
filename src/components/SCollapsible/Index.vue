<template>
  <div class="s-collapsible">
    <div class="header" @click="$emit('toggle', !isOpened)">
      <slot name="header">X</slot>
    </div>

    <div ref="wrapper" class="wrapper" :style="{ height: isOpened ? contentHeight + 'px' : 0 }">
      <slot />
    </div>
  </div>
</template>

<script>
export default {
  name: 'SCollapsible',

  props: {
    isOpened: { type: Boolean, required: true }
  },

  data: () => ({ observer: null, contentHeight: 0 }),

  mounted () {
    const target = this.$refs.wrapper.children[0]
    const config = { childList: true, subtree: true }
    const callback = elements => elements.forEach(({ _, target }) => this.contentHeight = target.clientHeight)

    this.observer = new MutationObserver(callback)
    this.observer.observe(target, config)

    this.contentHeight = target.clientHeight
  },

  beforeDestroy () {
    if (this.observer) this.observer.disconnect()
  }
}
</script>

<style lang="scss">
.s-collapsible {

  & > .wrapper {
    overflow: hidden;
    transition: height .3s ease-in-out;
  }
}
</style>
