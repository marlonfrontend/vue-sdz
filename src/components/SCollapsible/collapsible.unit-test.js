import { mount } from '@vue/test-utils'

import SCollapsible from './Index.vue'

/* eslint-disable */
describe('SCollapsible', () => {
  let cmp;

  beforeEach(() => {
    cmp = mount(SCollapsible, {
      propsData: {
        isOpened: true
      },
      slots: {
        header:'Header do collapsible',
        default: `Texto do collapsible`
      }
    });
  });


  test('is instantiated', () => {
    expect(cmp.exists()).toBeTruthy()
  })
  

  test("has the expected html structure", () => {
    expect(cmp.element).toMatchSnapshot();
  })

  test('check header text', () => {

    const text = cmp.find('div > div')
    expect(text.exists()).toBe(true)
    expect(text.text()).toBe('Header do collapsible')

  })

  test('check body text', () => {

    const text = cmp.find('.wrapper')

    expect(text.attributes().style).toContain('opacity')
    expect(text.exists()).toBe(true)
    expect(text.text()).toBe('Texto do collapsible')

  })

})
