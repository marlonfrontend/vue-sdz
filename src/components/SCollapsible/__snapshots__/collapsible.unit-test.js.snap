// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCollapsible has the expected html structure 1`] = `
<div
  class="s-collapsible"
>
  <div
    class="header"
  >
    Header do collapsible
  </div>
   
  <div
    class="wrapper"
    style="opacity: 1; height: 0px;"
  >
    Texto do collapsible
  </div>
</div>
`;
